ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,341 @@
+*)
+type state = Direct | Engaged | Disengaged
+
+type channel = {
+    queue: t Lwt_bounded_stream.t
+  ; push: t option -> unit option
+  ; lock: Lwt_mutex.t (* lock for the states *)
+  ; mutable state: state
+}
+
+(*
+   Notes:
+   - uses Mtime.t to force usage of monotonic time
+   - This means that between runs (and reboots) cached stated is lost if not
+     persisted first.
+   IDEA: carryover: read contents of cache and "convert it" to the current run
+
+   TODO:
+     - Add startup step to convert existing content to new time
+     - Exponential backoff on xapi push error
+     - Limit error logging on xapi push error: once per downtime is enough
+ *)
+
+module Writer : sig
+  val with_cache :
+       direct:(t -> string -> (unit, exn) Lwt_result.t)
+    -> Types.Service.t
+    -> channel
+    -> ((t -> string -> unit Lwt.t) -> 'a Lwt.t)
+    -> 'a Lwt.t
+  (** [with_cache ~direct typ context] creates a cache for content of type [typ]. The
+      cache is writable through the function [context], which is provided a
+      writing function to persist to the cache.
+
+      Example:
+        Xapi_guard.Disk_cache.(Writer.with_cache Tpm (upload session_cache)
+        @@ fun write_tpm -> write_tpm (uuid, time) contents
+    *)
+end = struct
+  let mkdir_safe name perm =
+    Lwt.catch
+      (fun () -> Lwt_unix.mkdir name perm)
+      (function
+        | Unix.(Unix_error (EEXIST, _, _)) -> Lwt.return_unit | e -> raise e
+        )
+
+  let files_in_existing dir =
+    let create_dir = function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          let* () = mkdir_safe dir 0o700 in
+          Lwt.return []
+      | e ->
+          raise e
+    in
+    files_in dir ~otherwise:create_dir
+
+  let persist_to ~filename ~contents =
+    let flags = Unix.[O_WRONLY; O_CREAT; O_SYNC] in
+    let perm = 0o600 in
+    let persist out_chan = Lwt_io.write out_chan contents in
+    Lwt_io.with_file ~flags ~perm ~mode:Output filename persist
+
+  let write_contents ~direct root queue (uuid, now) contents =
+    let __FUN = __FUNCTION__ in
+
+    let tpm_dir = root // Uuidm.(to_string uuid) in
+    (* 1. Record existing requests in cache *)
+    let* outdated_contents = files_in_existing tpm_dir in
+
+    let filename = tpm_dir // (Mtime.to_uint64_ns now |> Int64.to_string) in
+    (* 2. Write new timestamped content to cache, atomically, if needed; and
+       notify the other side, if needed *)
+    let persist () = persist_to ~filename ~contents in
+    let push () =
+      match queue.push (Some (uuid, now)) with
+      | Some () ->
+          Lwt.return_unit
+      | None ->
+          (* Queue is full, change mode to ignore queue *)
+          queue.state <- Disengaged ;
+          Lwt.return_unit
+    in
+    let engage_and_push exn =
+      queue.state <- Engaged ;
+      let* () =
+        Logs_lwt.info (fun m ->
+            m "%s: Error on push. Reason: %s" __FUN (Printexc.to_string exn)
+        )
+      in
+      push ()
```

Comment:
No `persist` before the first push?

Reply:
It needs to persist, indeed. I've rearranged the code so it's harder to push without persisting